// xml-trueformat/dist/esm/index.mjs
// util/escape.ts
var XML_ESCAPES_BY_ESCAPED_VALUE = {
  "&quot;": '"',
  "&amp;": "&",
  "&lt;": "<",
  "&gt;": ">",
  "&apos;": "'"
};
var XML_ESCAPES_BY_UNESCAPED_VALUE = {
  '"': "&quot;",
  "&": "&amp;",
  "<": "&lt;",
  ">": "&gt;",
  "'": "&apos;"
};
function escapeXml(value) {
  return value.replace(/["'<>&]/g, (match) => XML_ESCAPES_BY_UNESCAPED_VALUE[match]);
}
function unescapeXml(value) {
  return value.replace(/&quot;|&amp;|&lt;|&gt;|&apos;/g, (match) => XML_ESCAPES_BY_ESCAPED_VALUE[match]);
}

// model/xmlAttribute.ts
class XmlAttribute {
  name;
  value;
  leadingWs;
  wsBeforeEqual;
  wsAfterEqual;
  quote;
  hasValue;
  constructor(name, value, leadingWs = " ", wsBeforeEqual = "", wsAfterEqual = "", quote = '"', hasValue = true) {
    this.name = name;
    this.value = value;
    this.leadingWs = leadingWs;
    this.wsBeforeEqual = wsBeforeEqual;
    this.wsAfterEqual = wsAfterEqual;
    this.quote = quote;
    this.hasValue = hasValue;
  }
  unescapeValue() {
    return unescapeXml(this.value);
  }
  setValue(unescapedValue) {
    this.value = escapeXml(unescapedValue);
  }
  toString() {
    return this.leadingWs + this.name + this.wsBeforeEqual + (this.hasValue ? "=" + this.wsAfterEqual + this.quote + this.value + this.quote : "");
  }
}
// model/xmlNode.ts
class XmlNode {
}

// model/xmlCData.ts
class XmlCData extends XmlNode {
  cdata;
  constructor(cdata) {
    super();
    this.cdata = cdata;
  }
  toString() {
    return `<![CDATA[${this.cdata}]]>`;
  }
}
// model/xmlComment.ts
class XmlComment extends XmlNode {
  comment;
  constructor(comment) {
    super();
    this.comment = comment;
  }
  toString() {
    return `<!--${this.comment}-->`;
  }
}
// model/xmlDoctype.ts
class XmlDoctype extends XmlNode {
  content;
  constructor(content) {
    super();
    this.content = content;
  }
  toString() {
    return `<!DOCTYPE ${this.content}>`;
  }
}
// model/xmlText.ts
class XmlText extends XmlNode {
  text;
  constructor(text) {
    super();
    this.text = text;
  }
  unescapeText() {
    return unescapeXml(this.text);
  }
  setText(unescapedText) {
    this.text = escapeXml(unescapedText);
  }
  isWhitespace() {
    return this.text.trim() === "";
  }
  toString() {
    return this.text;
  }
}

// model/xmlElement.ts
class XmlElement extends XmlNode {
  tagName;
  attributes;
  children;
  attrTrailingWs;
  selfClosing;
  closeTagWs;
  constructor(tagName, attributes = [], children = [], attrTrailingWs = "", selfClosing = false, closeTagWs = "") {
    super();
    this.tagName = tagName;
    this.attributes = attributes;
    this.children = children;
    this.attrTrailingWs = attrTrailingWs;
    this.selfClosing = selfClosing;
    this.closeTagWs = closeTagWs;
  }
  getElementsByName(name) {
    return this.children.filter((c) => c instanceof XmlElement && c.tagName === name);
  }
  getFirstElementByName(name) {
    for (const child of this.children) {
      if (child instanceof XmlElement && child.tagName === name) {
        return child;
      }
    }
    return;
  }
  addElement(child, options) {
    let index;
    if (options?.after) {
      index = this.children.indexOf(options.after) + 1;
    } else if (options?.before) {
      index = this.children.indexOf(options.before);
    } else {
      index = this.children.length;
    }
    while (index > 0 && this.children[index - 1] instanceof XmlText && this.children[index - 1].isWhitespace()) {
      index--;
    }
    const sibling = this.children.slice(0, index).reverse().find((c) => c instanceof XmlElement) ?? this.children.slice(index).find((c) => c instanceof XmlElement);
    if (sibling) {
      const siblingIndex = this.children.indexOf(sibling);
      const textBefore = siblingIndex > 0 && this.children[siblingIndex - 1] instanceof XmlText && this.children[siblingIndex - 1].isWhitespace() ? this.children[siblingIndex - 1] : undefined;
      const leadingWs = textBefore?.text ?? "";
      this.addChild(index, new XmlText(leadingWs), child);
    } else {
      this.addChild(index, child);
    }
  }
  addChild(index, ...children) {
    if (this.selfClosing) {
      this.selfClosing = false;
    }
    this.children.splice(index, 0, ...children);
    this.joinTextNodes();
  }
  getAttribute(name) {
    return this.attributes.find((attr) => attr.name === name);
  }
  getAttributeValue(name) {
    return this.getAttribute(name)?.unescapeValue();
  }
  setAttributeValue(name, unescapedValue) {
    const attr = this.getAttribute(name);
    if (attr) {
      attr.setValue(unescapedValue);
    } else {
      const copyFromAttribute = this.attributes.at(-1);
      this.attributes.push(new XmlAttribute(name, unescapedValue, copyFromAttribute?.leadingWs, copyFromAttribute?.wsBeforeEqual, copyFromAttribute?.wsAfterEqual, copyFromAttribute?.quote));
    }
  }
  toString() {
    let openTag = `<${this.tagName}`;
    for (const attr of this.attributes) {
      openTag += attr.toString();
    }
    openTag += this.attrTrailingWs;
    openTag += this.selfClosing ? "/>" : ">";
    let childrenStr = "";
    if (!this.selfClosing) {
      for (const child of this.children) {
        childrenStr += child.toString();
      }
    }
    const closeTag = this.selfClosing ? "" : `</${this.tagName}${this.closeTagWs}>`;
    return openTag + childrenStr + closeTag;
  }
  joinTextNodes() {
    for (let i = 0;i < this.children.length - 1; i++) {
      let child = this.children[i];
      if (child instanceof XmlText && child.text === "") {
        this.children.splice(i, 1);
        i--;
      }
      if (child instanceof XmlText && this.children[i + 1] instanceof XmlText) {
        child.text += this.children[i + 1].text;
        this.children.splice(i + 1, 1);
        i--;
      }
    }
  }
}

// model/xmlDocument.ts
class XmlDocument extends XmlNode {
  children;
  constructor(children) {
    super();
    this.children = children;
  }
  getRootElement() {
    const root = this.children.find((child) => child instanceof XmlElement);
    if (!root) {
      throw new Error("No root element found");
    }
    return root;
  }
  toString() {
    return this.children.map((child) => child.toString()).join("");
  }
}
// model/xmlProcessing.ts
class XmlProcessing extends XmlNode {
  target;
  wsAfterTarget;
  data;
  constructor(target, wsAfterTarget, data) {
    super();
    this.target = target;
    this.wsAfterTarget = wsAfterTarget;
    this.data = data;
  }
  toString() {
    return `<?${this.target}${this.wsAfterTarget}${this.data}?>`;
  }
}
// xmlParser.ts
class XmlParser {
  input;
  pos = 0;
  constructor(input) {
    this.input = input;
  }
  static parseFragment(xml) {
    return XmlParser.parse(`<root>${xml}</root>`).getRootElement().children;
  }
  static parse(xml) {
    const parser = new XmlParser(xml);
    const children = [];
    while (parser.pos < parser.input.length) {
      const node = parser.parseNode();
      if (node) {
        children.push(node);
      } else {
        break;
      }
    }
    return new XmlDocument(children);
  }
  parseNode() {
    if (this.peek() === "<") {
      if (this.input.startsWith("<!--", this.pos)) {
        return this.parseComment();
      } else if (this.input.startsWith("<![CDATA[", this.pos)) {
        return this.parseCData();
      } else if (this.input.startsWith("<!DOCTYPE", this.pos)) {
        return this.parseDoctype();
      } else if (this.input.startsWith("<?", this.pos)) {
        return this.parseProcessing();
      } else if (this.input.startsWith("</", this.pos)) {
        return null;
      } else {
        return this.parseElement();
      }
    } else {
      return this.parseText();
    }
  }
  parseText() {
    const start = this.pos;
    while (this.pos < this.input.length && this.peek() !== "<") {
      this.pos++;
    }
    const text = this.input.substring(start, this.pos);
    return new XmlText(text);
  }
  parseComment() {
    const start = this.pos;
    this.pos += 4;
    const end = this.input.indexOf("-->", this.pos);
    if (end === -1) {
      throw new Error(`Unterminated comment starting at position ${start}`);
    }
    const commentContent = this.input.substring(this.pos, end);
    this.pos = end + 3;
    return new XmlComment(commentContent);
  }
  parseProcessing() {
    const start = this.pos;
    this.pos += 2;
    const target = this.readName();
    const wsAfterTarget = this.readWhitespace();
    const end = this.input.indexOf("?>", this.pos);
    if (end === -1) {
      throw new Error(`Unterminated processing instruction starting at position ${start}`);
    }
    const data = this.input.substring(this.pos, end);
    this.pos = end + 2;
    return new XmlProcessing(target, wsAfterTarget, data);
  }
  parseCData() {
    const start = this.pos;
    this.pos += 9;
    const end = this.input.indexOf("]]>", this.pos);
    if (end === -1) {
      throw new Error(`Unterminated CDATA section starting at position ${start}`);
    }
    const cdataContent = this.input.substring(this.pos, end);
    this.pos = end + 3;
    return new XmlCData(cdataContent);
  }
  parseDoctype() {
    const start = this.pos;
    this.pos += 9;
    const end = this.input.indexOf(">", this.pos);
    if (end === -1) {
      throw new Error(`Unterminated DOCTYPE declaration starting at position ${start}`);
    }
    const content = this.input.substring(this.pos, end).trim();
    this.pos = end + 1;
    return new XmlDoctype(content);
  }
  parseElement() {
    const start = this.pos;
    if (this.peek() !== "<") {
      throw new Error(`Expected '<' at position ${this.pos}`);
    }
    this.pos++;
    const tagName = this.readName();
    if (tagName === "") {
      throw new Error(`Expected tag name at position ${this.pos}`);
    }
    const attributes = [];
    let attrTrailingWs = "";
    while (this.pos < this.input.length && !this.startsWithAny([">", "/>"])) {
      const leadingWs = this.readWhitespace();
      if (this.startsWithAny([">", "/>"])) {
        attrTrailingWs = leadingWs;
        break;
      }
      const attrName = this.readName();
      if (attrName === "") {
        throw new Error(`Expected attribute name at position ${this.pos}`);
      }
      let wsBeforeEqual = this.readWhitespace();
      let wsAfterEqual = "";
      let quote = '"';
      let value = "";
      let hasValue = true;
      if (this.peek() === "=") {
        this.pos++;
        wsAfterEqual = this.readWhitespace();
        quote = this.peek();
        if (quote === '"' || quote === "'") {
          this.pos++;
          const startVal = this.pos;
          const endVal = this.input.indexOf(quote, this.pos);
          if (endVal === -1) {
            throw new Error(`Unterminated attribute value for "${attrName}" starting at position ${startVal}`);
          }
          value = this.input.substring(startVal, endVal);
          this.pos = endVal + 1;
        } else {
          throw new Error(`Expected quote character at position ${this.pos} for attribute "${attrName}"`);
        }
      } else {
        hasValue = false;
        this.pos -= wsBeforeEqual.length;
        wsBeforeEqual = "";
      }
      attributes.push(new XmlAttribute(attrName, value, leadingWs, wsBeforeEqual, wsAfterEqual, quote === '"' || quote === "'" ? quote : '"', hasValue));
    }
    let selfClosing = false;
    if (this.startsWithAny(["/>"])) {
      selfClosing = true;
      this.pos += 2;
    } else if (this.peek() === ">") {
      this.pos++;
    } else {
      throw new Error(`Expected '>' or '/>' at position ${this.pos}`);
    }
    const children = [];
    let closeTagWs = "";
    if (!selfClosing) {
      while (this.pos < this.input.length) {
        if (this.input.startsWith(`</${tagName}`, this.pos)) {
          break;
        }
        const child = this.parseNode();
        if (child instanceof XmlDoctype) {
          throw new Error(`Unexpected node type XmlDoctype as child of <${tagName}> at position ${this.pos}`);
        }
        if (child) {
          children.push(child);
        } else {
          break;
        }
      }
      if (this.input.startsWith(`</${tagName}`, this.pos)) {
        this.pos += 2;
        const closingTagName = this.readName();
        if (closingTagName !== tagName) {
          throw new Error(`Mismatched closing tag at position ${this.pos}: expected </${tagName}> but found </${closingTagName}>`);
        }
        closeTagWs = this.readWhitespace();
        if (this.peek() === ">") {
          this.pos++;
        } else {
          throw new Error(`Expected '>' at end of closing tag for <${tagName}> at position ${this.pos}`);
        }
      } else {
        throw new Error(`Unterminated element <${tagName}> starting at position ${start}`);
      }
    }
    return new XmlElement(tagName, attributes, children, attrTrailingWs, selfClosing, closeTagWs);
  }
  readName() {
    const start = this.pos;
    while (this.pos < this.input.length && /[A-Za-z0-9_\-.:\[\]*()#@]/.test(this.input[this.pos])) {
      this.pos++;
    }
    return this.input.substring(start, this.pos);
  }
  readWhitespace() {
    const start = this.pos;
    while (this.pos < this.input.length && /\s/.test(this.input[this.pos])) {
      this.pos++;
    }
    return this.input.substring(start, this.pos);
  }
  startsWithAny(strings) {
    for (const s of strings) {
      if (this.input.startsWith(s, this.pos)) {
        return true;
      }
    }
    return false;
  }
  peek() {
    return this.input[this.pos];
  }
}
export {
  XmlText,
  XmlProcessing,
  XmlParser,
  XmlNode,
  XmlElement,
  XmlDocument,
  XmlDoctype,
  XmlComment,
  XmlCData,
  XmlAttribute
};
